### PR TITLE
Fix 'isRequestOriginAllowed' returns mixed results when regex is used

### LINKED ISSUE #151: isRequestOriginAllowed function returning random (invalid) results when Regex is used
### Prerequisites

- [X] I have written a descriptive issue title
- [X] I have searched existing issues to ensure the bug has not already been reported

### Fastify version

3.20.2

### Plugin version

6.0.2

### Node.js version

16.6.2

### Operating system

Linux

### Operating system version (i.e. 20.04, 11.3, 10)

PopOS 20.04

### Description

Every other request, the isRequestOriginAllowed function, will return an invalid result, in case a Regex is used.

The behaviour can be tracked back to [this line in the code](https://github.com/fastify/fastify-cors/blob/170d3e236ad1a2279e71460788853b5313d465b4/index.js#L213)

The .test function is called, although this function retains state in the regex object. E.g. on first invocation, it will go through the regex until it is matched, and return the result. On second invocation, it will go through the remainder of the regex and reset, returning a wrong result. Then on the third invocation it produces the right result again. This behaviour is also described on the [MDN page of the function](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/RegExp/test):
> As with exec() (or in combination with it), test() called multiple times on the same global regular expression instance will advance past the previous match.

### Steps to Reproduce

Initialize the plugin with a Regex or list of regexes:
```
fastify.register(require('fastify-cors'), {
    origin: new RegExp('floodtags.com', 'gi'),
    methods: ['OPTIONS', 'GET', 'POST'],
    credentials: true,
    allowedHeaders: ['Authorization'],
  });
```

Now every other request from e.g. `https://www.floodtags.com/` will not get the correct `Access-Control-Allow-Origin` header

### Expected Behavior

The 'isRequestOriginAllowed' should return the same result on every invocation.

This can be achieved by e.g. using the String.match function instead of Regex.test. I can submit a pull-request for this if needed, please let me know!

### PR DESCRIPTION
**Pull request by Tom-Brouwer**
Fixes #151

This pull-request contains the code update (Regex.test replaced by String.match function), as well as the update to the tests (The tests now use a global regex, and repeats the test two times to test consistency)

#### Test Output
```
> fastify-cors@6.0.2 test
> npm run lint && npm run unit && npm run typescript


> fastify-cors@6.0.2 lint
> npm run lint:standard && npm run lint:typescript


> fastify-cors@6.0.2 lint:standard
> standard


> fastify-cors@6.0.2 lint:typescript
> standard --parser @typescript-eslint/parser --plugin @typescript-eslint/eslint-plugin test/*.ts


> fastify-cors@6.0.2 unit
> tap test/*.test.js

 PASS  test/vary.test.js 1 OK 6.878ms
 PASS  test/preflight.test.js 51 OK 168.181ms
 PASS  test/cors.test.js 139 OK 236.792ms

                         
  🌈 SUMMARY RESULTS 🌈  
                         

Suites:   3 passed, 3 of 3 completed
Asserts:  191 passed, of 191
Time:     791.297ms
----------|---------|----------|---------|---------|-------------------
File      | % Stmts | % Branch | % Funcs | % Lines | Uncovered Line #s 
----------|---------|----------|---------|---------|-------------------
All files |     100 |      100 |     100 |     100 |                   
 index.js |     100 |      100 |     100 |     100 |                   
 vary.js  |     100 |      100 |     100 |     100 |                   
----------|---------|----------|---------|---------|-------------------

> fastify-cors@6.0.2 typescript
> tsd
```

#### Checklist
- [ x ] run `npm run test` and ~~`npm run benchmark`~~ -> _Note: There is no 'benchmark' script in package.json_
- [ x ] tests and/or benchmarks are included
- [ x ] documentation is changed or added
- [ x ] commit message and code follows the [Developer's Certification of Origin](https://github.com/fastify/.github/blob/master/CONTRIBUTING.md#developers-certificate-of-origin-11)
      and the [Code of conduct](https://github.com/fastify/.github/blob/master/CODE_OF_CONDUCT.md)

### CODE REVIEW

File: index.js [hunk 1]
```diff
@@ -210,7 +210,7 @@ function isRequestOriginAllowed (reqOrigin, allowedOrigin) {
   } else if (typeof allowedOrigin === 'string') {
     return reqOrigin === allowedOrigin
   } else if (allowedOrigin instanceof RegExp) {
-    return allowedOrigin.test(reqOrigin)
+    return Boolean(reqOrigin.match(allowedOrigin))
```
Comment (zekth) [inline index.js:213]: can't we just do?
```suggestion
    allowedOrigin.lastIndex=0
    return allowedOrigin.test(reqOrigin)
```

Comment (Tom-Brouwer) [reply index.js:213]: @zekth I don't know, I'm not sure about the exact implementation of Node.js async/thread behaviour. E.g. if two requests are triggered simultaneously, can we ensure that the following does not occur?:
1. First request is running .test function
2. Second request sets 'lastIndex' to 0
3. First requests finds match and updates lastIndex again
4. Second requests then uses the updated lastIndex

If we are sure the above cannot happen, and these lines are always executed synchronously without interruption, we can also use your solution since it's less invasive.

Please let me know what you think, then I'll update the commit...

Comment (zekth) [reply index.js:213]: This cannot happen as we are not working with multithreading but asynchronization. As those operations happen in the same scope there would not be any collision regarding the change of the regex.

On the other hand this shows a problem that might appear, which is passing a reference of the regex in the plugin like:
```js
const reg = /(example|other)\.com/gi
fastify.register(cors, { origin: reg })
```
In this case, if `reg` is mutated anywhere else in the app we'll have a collision. So in this case the only way to avoid this is to do something like
```js
if (allowedOrigin instanceof RegExp) {
    const testReg = RegExp(allowedOrigin.source, allowedOrigin.flags)
    return testReg.test(reqOrigin)
}
```

Comment (Tom-Brouwer) [reply index.js:213]: You are correct. I just tested the performance implications of three variants:
```javascript
const sourceRegex = new RegExp('example.com', 'gi');
const testString = 'https://www.example.com/';

function timeRegular() { // Takes about 33ms
  start = performance.now();
  for (let i=0; i < 1000000; i++) {
    sourceRegex.lastIndex = 0;
    sourceRegex.test(testString);
  }
  end = performance.now()
  return end - start;
}

function timeCopy() { // Takes ~ 160 ms
  start = performance.now();
  for (let i=0; i < 1000000; i++) {
    const testRegex = new RegExp(sourceRegex.source, sourceRegex.flags);
    testRegex.test(testString);
  }
  end = performance.now()
  return end - start;
}

function timeMatch() { // Takes ~100ms
  start = performance.now();
  for (let i=0; i < 1000000; i++) {
    testString.match(sourceRegex);
  }
  end = performance.now()
  return end - start;
}
```

The copying seems to add some overhead to the process (factor 5 roughly), but for a million checks it's only about 130ms extra.

I also checked [here](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/RegExp), and it seems like 'lastIndex' is the only editable attribute for a RegExp object.

If you think this additional overhead compares positively to being sure that 'lastIndex' is zero at the start, please let me know, I'll implement your last suggestion.

Personally I would go with your first suggestion (Which is also the fastest from the above tests), since I wouldn't even want to promote being able to count on the 'lastIndex' staying zero anywhere in the plugin, but that may be just my personal preference...

Comment (zekth) [reply index.js:213]: @climba03003 I would like to have your opinion on this.

Comment (climba03003) [reply index.js:213]: I will go for editing the `lastIndex`. If you think it may alter by someone in the middle of run. You can clone it once before passing down to the check. It should be in https://github.com/fastify/fastify-cors/pull/152/files#R101

**Comment by Fdawgs** [comment]
@Tom-Brouwer how's this going?

**Comment by Tom-Brouwer** [comment]
Hi @Fdawgs ,

Sorry, totally forgot about this. I've pushed an update in-line with the last comments (So implementing the 'lastIndex' solution). Please review.

Best regards,

Tom Brouwer

### Test Output
```

> fastify-cors@6.0.2 test
> npm run lint && npm run unit && npm run typescript


> fastify-cors@6.0.2 lint
> npm run lint:standard && npm run lint:typescript


> fastify-cors@6.0.2 lint:standard
> standard


> fastify-cors@6.0.2 lint:typescript
> standard --parser @typescript-eslint/parser --plugin @typescript-eslint/eslint-plugin test/*.ts


> fastify-cors@6.0.2 unit
> tap test/*.test.js

 PASS  test/vary.test.js 1 OK 5.985ms
 PASS  test/preflight.test.js 51 OK 248.744ms
 PASS  test/cors.test.js 139 OK 262.177ms

                         
  🌈 SUMMARY RESULTS 🌈  
                         

Suites:   3 passed, 3 of 3 completed
Asserts:  191 passed, of 191
Time:     1s
----------|---------|----------|---------|---------|-------------------
File      | % Stmts | % Branch | % Funcs | % Lines | Uncovered Line #s 
----------|---------|----------|---------|---------|-------------------
All files |     100 |      100 |     100 |     100 |                   
 index.js |     100 |      100 |     100 |     100 |                   
 vary.js  |     100 |      100 |     100 |     100 |                   
----------|---------|----------|---------|---------|-------------------

> fastify-cors@6.0.2 typescript
> tsd

```

**Review by climba03003 — APPROVED** [review]
LGTM.

**Review by Fdawgs — APPROVED** [review]
